### ISSUE TITLE
URI schemes are strange

### ISSUE THREAD

**Issue by dwmw2** [tpm2-software/tpm2-openssl#71]
At https://github.com/tpm2-software/tpm2-openssl/blob/1.2.0-rc1/src/tpm2-provider-store-object.c#L59 we seem to use `object:` as the URI scheme for... a file? Shouldn't that be `file:` ? 

Then at https://github.com/tpm2-software/tpm2-openssl/blob/1.2.0-rc1/src/tpm2-provider-store-object.c#L63 we use `handle:`. That's a really strange choice; shouldn't it have `tpm` somewhere in the scheme name?

See https://www.ietf.org/archive/id/draft-mavrogiannopoulos-tpmuri-01.txt for example, which defined a URI scheme for TPMv1.2 keys. Let's do this properly for TPMv2 too?

Tag @jejb

**Comment by gotthardp** [comment]
Well, everything can be improved. The `object` scheme was just easier to implement.

**Comment by jejb** [comment]
On Fri, 2023-05-19 at 04:01 -0700, David Woodhouse wrote:
> At
> https://github.com/tpm2-software/tpm2-openssl/blob/1.2.0-rc1/src/tpm2-provider-store-object.c#L59
>  we seem to use `object:` as the URI scheme for... a file? Shouldn't
> that be `file:` ? 

Technically it is.  The default file scheme, which every provider also
uses, has an optional file: prefix.

> Then at
> https://github.com/tpm2-software/tpm2-openssl/blob/1.2.0-rc1/src/tpm2-provider-store-object.c#L63
>  we use `handle:`. That's a really strange choice; shouldn't it have
> `tpm` somewhere in the scheme name?
> 
> See
> https://www.ietf.org/archive/id/draft-mavrogiannopoulos-tpmuri-01.txt
>  for example, which defined a URI scheme for TPMv1.2 keys. Let's do
> this properly for TPMv2 too?
> 
> Tag @jejb

You want to know what I use in my provider?  By default //nvkey for
backwards compatibility with the engine: 

https://git.kernel.org/pub/scm/linux/kernel/git/jejb/openssl_tpm2_engine.git/tree/src/provider/store.c#n14

But it's also settable to anything in the config file.

James

**Comment by dwmw2** [comment]
Agreed. *Files* have no place in such a URI format because we have `file:` URIs for that, and the standard PEM form that we've already agreed on covered by your doc at : https://www.hansenpartnership.com/draft-bottomley-tpm2-keys.html

The other parts of that older draft made sense for TPMv1.2 but things are different for TPMv2 so a new definition is needed. I was merely presenting it as an example of the genre.

**Comment by AndreasFuchsTPM** [comment]
FWIW my 2 cents:
- file: for files
- handle: for persistent (and transient; even though irrelevant) handles
- tss2: for TSS2 objects (e.g. tss2:/nv/owner/mykey)
- no-prefix:
  - if starts with 0x then handle:
  - else file:


I agree that object: is weird.

Proposal:
- Let's agree on a scheme for TPM 2.0 (let's discard the TPM 1.2 scheme)
- I propose to agree on aforementioned scheme
- Let's keep object: in this project but mark it as deprecated

**Comment by dwmw2** [comment]
I finally got round to updating http://david.woodhou.se/draft-woodhouse-cert-best-practice.html to include TPMv2 support. It references James's ASN.1 draft but laments the lack of a URI form for TPMv2 as I didn't think I could use https://github.com/tpm2-software/tpm2-openssl/issues/71#issuecomment-1560914554 as a normative reference. 

I'd love to see that defined *somewhere* I can reference it, even if it's just an extra one-line section in James's doc?

**Comment by dwmw2** [comment]
I'd love to take both https://datatracker.ietf.org/doc/draft-woodhouse-cert-best-practice/ and @jejb's draft (James, did you ever submit it? I thought you had but I can't find it now) through.

I do think we need to define TPMv2 URIs for keys in persistent storage _(I think TPMv1.2 can probably die, or we can subsume @nmav's old draft if we really want to)_. I have a slight preference for doing that as *part* of James's doc, but ISTR last time we discussed it James did not, so I suppose it could be another separate doc.
